In [135]:
import numpy as np
import pandas as pd

In [136]:
books = pd.read_csv("Books.csv")
users = pd.read_csv("users.csv")
ratings = pd.read_csv("ratings.csv")

C:\Users\NutanTripathy\AppData\Local\Temp\ipykernel_3528\3358372675.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv("Books.csv")


In [137]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [138]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

The above dataset is used to store user data ( their userid, location and age)

In [139]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

So above dataset has 3 columns, column 1 is userid(specific id given to the user), column 2 is ISBN number (unique number given to the books) and book rating is the rating given by a specific user to a specific book

In [140]:
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 8)
(1149780, 3)
(278858, 3)


In [141]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [142]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [143]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [144]:
books.duplicated().sum()

np.int64(0)

In [145]:
ratings.duplicated().sum()

np.int64(0)

In [146]:
users.duplicated().sum()

np.int64(0)

# Popularity Based Recommender System

In [147]:
ratings_with_name = ratings.merge(books,on='ISBN')

In [148]:
num_rating_df = ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating':'num_ratings'},inplace=True)
num_rating_df


Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [149]:
avg_rating_df = ratings_with_name.groupby('Book-Title')['Book-Rating'].mean().reset_index()
avg_rating_df.rename(columns={'Book-Rating':'avg_rating'},inplace=True)
avg_rating_df

Book-Title  avg_rating
0        A Light in the Storm: The Civil War Diary of ...    2.250000
1                                   Always Have Popsicles    0.000000
2                    Apple Magic (The Collector's series)    0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...    8.000000
4        Beyond IBM: Leadership Marketing and Finance ...    0.000000
...                                                   ...         ...
241066                                      Ã?Â?lpiraten.    0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.    5.250000
241068                                Ã?Â?sterlich leben.    7.000000
241069                              Ã?Â?stlich der Berge.    2.666667
241070                                  Ã?Â?thique en toc    4.000000

[241071 rows x 2 columns]

Now the popular df will be given by 

In [150]:

popular_df = num_rating_df.merge(avg_rating_df,on='Book-Title')
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_rating  
0         2.250000  
1         0.000000  
2         0.000000  
3         8.000000  
4         0.000000  
...            ...  
241066    0.000000  
241067    5.250000  
241068    7.000000  
241069    2.666667  
241070    4.000000  

[241071 rows x 3 columns]

In [151]:
popular_df = popular_df[popular_df["num_ratings"]>=250].sort_values("avg_rating",ascending=False)

In [152]:
popular_df

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
...                                                   ...          ...   
227701       Vinegar Hill (Oprah's Book Club (Paperback))          265   
233635                                           Whispers          286   
143377                                  Presumed Innocent          294   
94382                                        Isle of Dogs          288   
234951                                        Wild Animus         2502   

        avg_rating  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453  
...            ...  
227701    2.245283  
233635    2.199301  
143377    2.139456  
94382     2.000000  
234951    1.019584  

[186 rows x 3 columns]

merge in python acts as join in sql

to get the name of book author, image related to it, year of publication and other details, we will join popular_df with book df and our common column will be Book title

In [153]:
popular_df.merge(books,on="Book-Title").drop_duplicates("Book-Title")[["Book-Title","Book-Author","num_ratings","avg_rating"]]

Book-Title        Book-Author  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)      J. K. Rowling   
3         Harry Potter and the Goblet of Fire (Book 4)      J. K. Rowling   
5       Harry Potter and the Sorcerer's Stone (Book 1)      J. K. Rowling   
9    Harry Potter and the Order of the Phoenix (Boo...      J. K. Rowling   
13    Harry Potter and the Chamber of Secrets (Book 2)      J. K. Rowling   
..                                                 ...                ...   
716       Vinegar Hill (Oprah's Book Club (Paperback))   A. Manette Ansay   
717                                           Whispers        BELVA PLAIN   
727                                  Presumed Innocent        Scott Turow   
733                                       Isle of Dogs  Patricia Cornwell   
739                                        Wild Animus       Rich Shapero   

     num_ratings  avg_rating  
0            428    5.852804  
3            387    5.824289  
5            278    5.737410  
9            347    5.501441  
13           556    5.183453  
..           ...         ...  
716          265    2.245283  
717          286    2.199301  
727          294    2.139456  
733          288    2.000000  
739         2502    1.019584  

[186 rows x 4 columns]

# Collaborative Filtering Based Recommender System

Here we will also consider, how many people are serious readors and we will consider the feedback of those serious readors to reccomend a book to another user. 

Meanwhile we also need to consider which books are popular, how many ratings they are getting 

In [154]:
ratings_with_name

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1         276726  0155061224            5   
2         276727  0446520802            0   
3         276729  052165615X            3   
4         276729  0521795028            6   
...          ...         ...          ...   
1031131   276704  0876044011            0   
1031132   276704  1563526298            9   
1031133   276706  0679447156            0   
1031134   276709  0515107662           10   
1031135   276721  0590442449           10   

                                                Book-Title        Book-Author  \
0                                     Flesh Tones: A Novel         M. J. Rose   
1                                         Rites of Passage         Judith Rae   
2                                             The Notebook    Nicholas Sparks   
3                                           Help!: Level 1      Philip Prowse   
4        The Amsterdam Connection : Level 4 (Cambridge ...        Sue Leather   
...                                                    ...                ...   
1031131  Edgar Cayce on the Akashic Records: The Book o...  Kevin J. Todeschi   
1031132  Get Clark Smart : The Ultimate Guide for the S...       Clark Howard   
1031133  Eight Weeks to Optimum Health: A Proven Progra...        Andrew Weil   
1031134   The Sherbrooke Bride (Bride Trilogy (Paperback))  Catherine Coulter   
1031135                                  Fourth Grade Rats     Jerry Spinelli   

        Year-Of-Publication  \
0                      2002   
1                      2001   
2                      1996   
3                      1999   
4                      2001   
...                     ...   
1031131                1998   
1031132                2000   
1031133                1997   
1031134                1996   
1031135                1996   

                                                 Publisher  \
0                                         Ballantine Books   
1                                                   Heinle   
2                                             Warner Books   
3                               Cambridge University Press   
4                               Cambridge University Press   
...                                                    ...   
1031131  A.R.E. Press (Association of Research &amp; Enlig   
1031132                                   Longstreet Press   
1031133                                    Alfred A. Knopf   
1031134                                         Jove Books   
1031135                                         Scholastic   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0876044011.0...   
1031132  http://images.amazon.com/images/P/1563526298.0...   
1031133  http://images.amazon.com/images/P/0679447156.0...   
1031134  http://images.amazon.com/images/P/0515107662.0...   
1031135  http://images.amazon.com/images/P/0590442449.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0876044011.0...   
1031132  http://images.amazon.com/images/P/1563526298.0...   
1031133  http://images.amazon.com/images/P/0679447156.0...   
1031134  htt

In [155]:
#Here we are checking the number of users who have given more than 200 ratings
users = ratings_with_name.groupby("User-ID").count()["Book-Rating"] > 200

serious_users = users[users].index

In [158]:
filtered_users = ratings_with_name[ratings_with_name["User-ID"].isin(serious_users)]
filtered_users

User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
1151      277427  0026217457            0   
1152      277427  003008685X            8   
1153      277427  0030615321            0   
1154      277427  0060002050            0   
...          ...         ...          ...   
1029357   275970  1931868123            0   
1029358   275970  3411086211           10   
1029359   275970  3829021860            0   
1029360   275970  4770019572            0   
1029361   275970  9626340762            8   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1151                    Vegetarian Times Complete Cookbook   
1152                                              Pioneers   
1153      Ask for May, Settle for June (A Doonesbury book)   
1154                     On a Wicked Dawn (Cynster Novels)   
...                                                    ...   
1029357  There's a Porcupine in My Outhouse: Misadventu...   
1029358                                         Die Biene.   
1029359                                     The Penis Book   
1029360                                            Musashi   
1029361  Northanger Abbey (Classic Literature with Clas...   

                           Book-Author Year-Of-Publication  \
1150                 James Finn Garner                1994   
1151                        Lucy  Moll                1995   
1152             James Fenimore Cooper                1974   
1153                     G. B. Trudeau                1982   
1154                 Stephanie Laurens                2002   
...                                ...                 ...   
1029357                   Mike Tougias                2002   
1029358  Sybil Gr�¤fin Sch�¶nfeldt                1993   
1029359                   Joseph Cohen                1999   
1029360                 Eiji Yoshikawa                1995   
1029361                    Jane Austen                1996   

                                    Publisher  \
1150                John Wiley &amp; Sons Inc   
1151                    John Wiley &amp; Sons   
1152                         Thomson Learning   
1153                      Henry Holt &amp; Co   
1154                               Avon Books   
...                                       ...   
1029357                    Capital Books (VA)   
1029358  Bibliographisches Institut, Mannheim   
1029359                              Konemann   
1029360          Kodansha International (JPN)   
1029361                 Naxos Audiobooks Ltd.   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.amazon.com/images/P/4770019572.0...   
1029361  http://images.amazon.com/images/P/9626340762.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.a

So 474007 ratings are given by just these top 200 serious users

In [161]:
famous_books = filtered_users.groupby("Book-Title").count()["Book-Rating"] > 50
famous_books

Book-Title
 A Light in the Storm: The Civil War Diary of Amelia Martin, Fenwick Island, Delaware, 1861 (Dear America)    False
 Always Have Popsicles                                                                                        False
 Apple Magic (The Collector's series)                                                                         False
 Beyond IBM: Leadership Marketing and Finance for the 1990s                                                   False
 Clifford Visita El Hospital (Clifford El Gran Perro Colorado)                                                False
                                                                                                              ...  
Ã?Â?ber das Fernsehen.                                                                                        False
Ã?Â?ber die Pflicht zum Ungehorsam gegen den Staat.                                                           False
Ã?Â?lpiraten.                                                

In [162]:
famous_books = famous_books[famous_books].index

In [163]:
famous_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=679)

In [165]:
final_ratings = filtered_users[filtered_users["Book-Title"].isin(famous_books)]

In [166]:
final_ratings.drop_duplicates()

User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
1163      277427  0060930535            0   
1165      277427  0060934417            0   
1168      277427  0061009059            9   
1174      277427  006440188X            0   
...          ...         ...          ...   
1029196   275970  1400031354            0   
1029197   275970  1400031362            0   
1029270   275970  1573229725            0   
1029309   275970  1586210661            9   
1029310   275970  1586212230            0   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1163                         The Poisonwood Bible: A Novel   
1165                                    Bel Canto: A Novel   
1168     One for the Money (Stephanie Plum Novels (Pape...   
1174                                     The Secret Garden   
...                                                    ...   
1029196  Tears of the Giraffe (No.1 Ladies Detective Ag...   
1029197  Morality for Beautiful Girls (No.1 Ladies Dete...   
1029270                                        Fingersmith   
1029309                             Me Talk Pretty One Day   
1029310                                              Naked   

                     Book-Author Year-Of-Publication  \
1150           James Finn Garner                1994   
1163          Barbara Kingsolver                1999   
1165                Ann Patchett                2002   
1168             Janet Evanovich                1995   
1174     Frances Hodgson Burnett                1998   
...                          ...                 ...   
1029196   Alexander McCall Smith                2002   
1029197   Alexander McCall Smith                2002   
1029270             Sarah Waters                2002   
1029309            David Sedaris                2001   
1029310            David Sedaris                2001   

                         Publisher  \
1150     John Wiley &amp; Sons Inc   
1163                     Perennial   
1165                     Perennial   
1168                   HarperTorch   
1174                  HarperTrophy   
...                            ...   
1029196                     Anchor   
1029197                     Anchor   
1029270            Riverhead Books   
1029309    Time Warner Audio Major   
1029310    Time Warner Audio Major   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1163     http://images.amazon.com/images/P/0060930535.0...   
1165     http://images.amazon.com/images/P/0060934417.0...   
1168     http://images.amazon.com/images/P/0061009059.0...   
1174     http://images.amazon.com/images/P/006440188X.0...   
...                                                    ...   
1029196  http://images.amazon.com/images/P/1400031354.0...   
1029197  http://images.amazon.com/images/P/1400031362.0...   
1029270  http://images.amazon.com/images/P/1573229725.0...   
1029309  http://images.amazon.com/images/P/1586210661.0...   
1029310  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1163     http://images.amazon.com/images/P/0060930535.0...   
1165     http://images.amazon.com/images/P/0060934417.0...   
1168     http://images.amazon.com/images/P/0061009059.0...   
1174     http://images.amazon.com/images/P/006440188X.0...   
...                                                    ...   
1029196  http://images.amazon.com/images/P/1400031354.0...   
1029197  http://images.amazon.com/images/P/1400031362.0...   
1029270  http://images.amazon.com/images/P/1573229725.0...   
1029309  http://images.amazon.com/images/P/1586210661.0...   
1029310  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-L  
1150     http://images.amazon.com/images/P

In [168]:
pt = final_ratings.pivot_table(index="Book-Title",columns="User-ID",values="Book-Rating")

In [169]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [170]:
pt.fillna(0,inplace=True)

In [171]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

in above pivot table, each book can be seen as a vector in a 679 dimensional space. Now we need to find how these vectors are similar to each other by calculating Euclidien distance between them. 

In [174]:
from sklearn.metrics.pairwise import cosine_similarity

In [180]:
similarity_score = cosine_similarity(pt)

Similarity score is signifying the similarity between each book eith each of the book

In [185]:
print(list(enumerate(similarity_score[0])))

[(0, np.float64(0.9999999999999998)), (1, np.float64(0.10255024829874466)), (2, np.float64(0.012208555646509366)), (3, np.float64(0.0)), (4, np.float64(0.053672244225442485)), (5, np.float64(0.027749013024662858)), (6, np.float64(0.08216491130000354)), (7, np.float64(0.13732869060263378)), (8, np.float64(0.03261686350494277)), (9, np.float64(0.03667590746512177)), (10, np.float64(0.02322418425912186)), (11, np.float64(0.06766486753662343)), (12, np.float64(0.020839782065795794)), (13, np.float64(0.09673734976610586)), (14, np.float64(0.0830311169181103)), (15, np.float64(0.1115354275858341)), (16, np.float64(0.05100410510067514)), (17, np.float64(0.025177841632809915)), (18, np.float64(0.11706383185253948)), (19, np.float64(0.0)), (20, np.float64(0.14333792815336935)), (21, np.float64(0.07847534327232125)), (22, np.float64(0.06150451339525138)), (23, np.float64(0.08723968326958673)), (24, np.float64(0.0)), (25, np.float64(0.07009814362931202)), (26, np.float64(0.13658680855472938)), (2

In [187]:
def recommend(book_name):
    #fetching the index of the book
    index = np.where(pt.index==book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1], reverse=True)[1:6]

    suggestions = []
    for i in similar_items:
        print(pt.index[i[0]])
        # suggestions.append(pt.index[i[0]])

    # return suggestions

In [188]:
recommend("Message in a Bottle")

Nights in Rodanthe
The Mulberry Tree
A Walk to Remember
River's End
Nightmares &amp; Dreamscapes


In [189]:
recommend("1984")

Animal Farm
The Handmaid's Tale
Brave New World
The Vampire Lestat (Vampire Chronicles, Book II)
The Hours : A Novel
